### PR TITLE
Angelina nedelcu assignment3

### PR DESCRIPTION
**Pull request by angelina-nedelcu**
Solution for assignment 3 with bonus no. 1

### CODE REVIEW

File: angelina-nedelcu/assignment-3/WordSearchHelper.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	 * Recursive function. Base case is when a letter does not have neighbors that create
+	 * a valid prefix;
+	 * For a given position, check all possible neighbors to see if they form together with
+	 * the current prefix a word or a prefix. If it is a word it is added in the final list.
+	 * If it is a valid prefix, then a recursive call is launched.
+	 * 
+	 * Worst case space complexity: O (length * width), where length * width are the dimensions 
+	 * of the grid; this is because of the recursive call;
+	 * 
+	 * Worst case time complexity : O (length * width * nr_neighbors * O(isWord)))
```
Comment (imorlowska) [inline angelina-nedelcu/assignment-3/WordSearchHelper.java]: Are you sure? :)

File: angelina-nedelcu/assignment-3/WordSearchHelperTester.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+		lexicon.addWord("CAT");
+		lexicon.addWord("CART");
+		
+		/*
+		 * The matrix is:
+		 * A A R
+		 * T C D
+		 */
+		
+		Grid grid = new Grid(2, 3);
```
Comment (imorlowska) [inline angelina-nedelcu/assignment-3/WordSearchHelperTester.java]: You can create a new char[][] array literal and then pass that to the Grid's constructor - to make it easier to write and more readable too. E.g.:
```
char[][] input = {
    {'A', 'A', "R'},
    {'T', 'C', 'D'}
};
```

File: angelina-nedelcu/assignment-3/ArrayLexicon.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+		return allWords.contains(potentialWord);
+	}
+
+	/* 
+	 * Worst time complexity : O(word_number * size_longest_word)
+	 */
+	@Override
+	public boolean isPrefix(String potentialPrefix) {
+		for (String word : allWords) {
+			if (word.startsWith(potentialPrefix) && word.length() > potentialPrefix.length()) {
```
[the remark this answers is not in the document]
Comment (angelina-nedelcu) [reply angelina-nedelcu/assignment-3/ArrayLexicon.java]: I think that I should store for every word its length and avoid calculating the lengths more than once. Also, I should store in an integer the prefix length too when creating it.

File: angelina-nedelcu/assignment-3/Grid.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+
+public class Grid {
+
+	private Character[][] lettersGrid;
```
[the remark this answers is not in the document]
Comment (angelina-nedelcu) [reply angelina-nedelcu/assignment-3/Grid.java]: You are right! Thanks!

File: angelina-nedelcu/assignment-3/Grid.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	private int rowsNumber;
+	private int colsNumber; 
+	
+	public Grid(int rowsNumber, int colsNumber) {
```
[the remark this answers is not in the document]
Comment (angelina-nedelcu) [reply angelina-nedelcu/assignment-3/Grid.java]: It is a great idea!

File: angelina-nedelcu/assignment-3/TrieLexiconTester.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+	@Test
+	public void testPrefixNotExiting() {
+		TrieLexicon trieLexicon = new TrieLexicon();
+
+		trieLexicon.addWord("ANA");
+		trieLexicon.addWord("ANB");
+		trieLexicon.addWord("CAR");
+		assertFalse(trieLexicon.isPrefix("BAN"));
+	}
+
```
[the remark this answers is not in the document]
Comment (angelina-nedelcu) [reply angelina-nedelcu/assignment-3/TrieLexiconTester.java]: Solved!

File: angelina-nedelcu/assignment-3/WordSearchHelper.java (hunk 1 above)
Comment (angelina-nedelcu) [reply angelina-nedelcu/assignment-3/WordSearchHelper.java]: I have updated in the new files.

File: angelina-nedelcu/assignment-3/WordSearchHelperTester.java (hunk 2 above)
Comment (angelina-nedelcu) [reply angelina-nedelcu/assignment-3/WordSearchHelperTester.java]: Solved! Thanks!

**Pull request merged**